pkg/report/reporter.go

```diff
@@ -0,0 +1,199 @@
+	v := ctx.Value(reporterContextKey{})
+	if v == nil {
+		return &discardReporter{}
+	}
+	switch v := v.(type) {
+	case *defaultReporter:
+		return v
+	default:
+		panic(fmt.Sprintf("unexpected value type for reporter context key: %T", v))
+	}
+}
+
+type Reporter interface {
+	ReportDeployment(config coordinate.Coordinate, state string, details []Detail, err error)
+	GetSummary() string
+	Stop()
+}
+
+type Clock interface {
+	Now() time.Time
+}
+
+type rtcClock struct{}
+
+func (c *rtcClock) Now() time.Time {
+	return time.Now()
+}
+
+type defaultReporter struct {
+	queue                    chan Record
+	mu                       sync.Mutex
+	wg                       sync.WaitGroup
+	clock                    Clock
+	started                  time.Time
+	ended                    time.Time
+	deploymentsSuccessCount  int
+	deploymentsErrorCount    int
+	deploymentsExcludedCount int
+	deploymentsSkippedCount  int
+}
+
+func NewDefaultReporter(fs afero.Fs, reportFilePath string) Reporter {
+	return NewDefaultReporterWithClock(fs, reportFilePath, &rtcClock{})
+}
+
+func NewDefaultReporterWithClock(fs afero.Fs, reportFilePath string, c Clock) Reporter {
+	r := &defaultReporter{
+		clock:   c,
+		started: c.Now(),
+		queue:   make(chan Record, 32),
+	}
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+		if err := r.runRecorder(fs, reportFilePath); err != nil {
+			log.Error("Error recording deployment report: %s", err)
+		}
+	}()
+	return r
+}
+
+func (d *defaultReporter) runRecorder(fs afero.Fs, reportFilePath string) error {
+	file, err := fs.OpenFile(reportFilePath, os.O_TRUNC|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return fmt.Errorf("error open record file: %w", err)
+	}
+
+	writer := bufio.NewWriter(file)
+	for r := range d.queue {
+		d.updateSummaryFromRecord(r)
+
+		b, err := json.Marshal(r)
+		if err != nil {
+			return fmt.Errorf("unable to convert record: %w", err)
+		}
+
+		if _, err := writer.Write(b); err != nil {
+			return fmt.Errorf("unable to write record: %w", err)
+		}
+
+		if _, err := writer.WriteString("\n"); err != nil {
+			return fmt.Errorf("unable to write newline: %w", err)
+		}
+	}
+
+	if err := writer.Flush(); err != nil {
+		return fmt.Errorf("unable to flush record file: %w", err)
+	}
+
+	if err := file.Close(); err != nil {
+		return fmt.Errorf("unable to close record file: %w", err)
+	}
+	return nil
+}
+
+func (d *defaultReporter) updateSummaryFromRecord(r Record) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	d.ended = time.Time(r.Time)
+	switch r.State {
+	case State_DEPL_SUCCESS:
+		d.deploymentsSuccessCount++
+	case State_DEPL_EXCLUDED:
+		d.deploymentsExcludedCount++
+	case State_DEPL_SKIPPED:
+		d.deploymentsSkippedCount++
+	case State_DEPL_ERR:
+		d.deploymentsErrorCount++
+	default:
+		panic(fmt.Sprintf("unexpected state for deployment event: %s", r.State))
+	}
+}
+
+func (d *defaultReporter) ReportDeployment(config coordinate.Coordinate, state string, details []Detail, err error) {
+	d.queue <- Record{
```

Comment:
What do you think about haveing more than one type of `Record`?
And than `reportDeployment` would accept interface `Record`. With that I think you woud achive beter modularity

Reply:
Ultimately this is true, depends on the requirements for other sorts of events. I did experiment with this, but until we know what is needed it just adds extra complexity.